esp_delta_ota/examples/https_delta_ota/tools/esp_delta_ota_patch_gen.py

```diff
@@ -19,40 +23,120 @@
 # Magic Byte is created using command: echo -n "esp_delta_ota" | sha256sum
 esp_delta_ota_magic = 0xfccdde10
 
-MAGIC_SIZE = 4
-DIGEST_SIZE = 32
-RESERVED_HEADER = 64 - (MAGIC_SIZE + DIGEST_SIZE)
+MAGIC_SIZE = 4 # This is size of magic byte
+DIGEST_SIZE = 32 # This is SHA256 of the base binary    
+HEADER_VERSION_SIZE = 1 # This is size of header version
+CRC_SIZE = 4 # This is CRC32 of patch file which shall be applied on the base binary
+HEADER_SIZE = 64
+RESERVED_HEADER = HEADER_SIZE - (MAGIC_SIZE + DIGEST_SIZE + HEADER_VERSION_SIZE + CRC_SIZE) # This is reserved header size
+
+HEADER_VERSION = 0 # Define the HEADER_VERSION as 0
+
+def calculate_sha256(file_path: str) -> str:
+    """Calculate the SHA-256 hash of a file."""
+    sha256_hash = hashlib.sha256()
+    
+    with open(file_path, "rb") as f:
+        # Read the file in chunks to avoid memory issues for large files
+        for byte_block in iter(lambda: f.read(4096), b""):
+            sha256_hash.update(byte_block)
+    
+    # Return the hex representation of the hash
+    return sha256_hash.hexdigest()
 
 def create_patch(chip: str, base_binary: str, new_binary: str, patch_file_name: str) -> None:
     cmd = "esptool.py --chip " + chip + " image_info " + base_binary
     proc = subprocess.Popen([cmd], stdout=subprocess.PIPE, shell=True)
     (out, err) = proc.communicate()
-    x = re.search(b"Validation Hash: ([A-Za-z0-9]+) \(valid\)", out)
+    x = re.search(b"Validation Hash: ([A-Za-z0-9]+) \\(valid\\)", out)
 
     os.system("detools create_patch -c heatshrink " + base_binary + " " + new_binary + " " + patch_file_name)
     patch_file_without_header = "patch_file_temp.bin"
     os.system("mv " + patch_file_name + " " + patch_file_without_header)
 
+    with open(patch_file_without_header, "rb") as temp_patch:
+        patch_content = temp_patch.read()
+        crc32 = zlib.crc32(patch_content)
+
     with open(patch_file_name, "wb") as patch_file:
         patch_file.write(esp_delta_ota_magic.to_bytes(MAGIC_SIZE, 'little'))
         patch_file.write(bytes.fromhex(x[1].decode()))
+        patch_file.write(HEADER_VERSION.to_bytes(HEADER_VERSION_SIZE, 'little'))
```

Comment:
This now changes the image format.

- Do we need to update the image format diagram in the component?
- Should we update the example to handle the CRC verification part?
- How can we make this change backward compatible? Earlier format won't have the CRC field.

Reply:
we have removed the CRC field from the header as it is not possible to calculate the base binary CRC on device, because detools applies the patch in chunk i.e. unable to get the patch size at the beginning.